service/replay.go

```diff
@@ -100,3 +100,44 @@ func (c *ReplayCache) Add(id string, salt []byte) bool {
 	c.active[hash] = empty{}
 	return !inArchive
 }
+
+// Resize adjusts the capacity of the ReplayCache.
+//
+// If the new capacity is less than the current capacity, and the number of
+// active handshakes exceeds the new capacity, then we move the excess to the
+// archive.
+func (c *ReplayCache) Resize(capacity int) {
+	if capacity > MaxCapacity {
+		panic("ReplayCache capacity would result in too many false positives")
+	}
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// Shrink the active handshakes to capacity and move the rest to the archive.
+	if capacity < c.capacity {
+		active := make(map[uint32]empty, capacity)
+		archive := make(map[uint32]empty, 0)
+
+		// Move handshakes up to capacity into a new active, and the remainder into a new archive.
+		for k, v := range c.active {
+			if len(active) != capacity {
+				active[k] = v
+			} else if len(archive) != capacity {
```

Comment:
The archive here can fill up to the old capacity.
Also, the ones you keep in active are arbitrary, not in sequence order.
So this assignment is a bit unnecessary. It can also be a lot of work that may hold the server if the map is large. You may see CPU spikes on config updates.

I think it's fine to just not change anything other than setting the new capacity.

Eventually (or immediately) `active` will hit the capacity, triggering the allocation of a new map with the new capacity, and later the `archive` will be gone as well and both will have the new capacity.